File: Chapter 4 Comparison of upper bounds/computeCombettes.m
function theta_m = Combettes(weights,net_dim, alpha)
    
    hidden_layer_dim = net_dim(2:end-1);
    n_hidden_neurons = sum(net_dim(2:end-1));
    n_hidden_layers = length(net_dim) -2;
    val = 1-2*alpha;

    N = n_hidden_neurons;
    comb_matrix = dec2bin(0:2^N-1)' - '0';
    
    theta_m = 0;
    for k = 1:2^n_hidden_neurons
        pattern = comb_matrix(:,k);  

        last_value = 1;
        W_ges = weights{1};
        
        for n_layer = 1: n_hidden_layers
            pattern_part1 = pattern(last_value:last_value+hidden_layer_dim(n_layer)-1);
        
            pattern_part2 = ones(hidden_layer_dim(n_layer),1) - pattern_part1;
            
            pattern_part = val*pattern_part1 + pattern_part2;
            
            Lambda = diag(pattern_part);
            last_value = last_value + hidden_layer_dim(n_layer);
            W_ges = Lambda*W_ges;
            W_next = weights{n_layer +1};
            W_ges = W_next*W_ges;
        end

        Norm_linear_subregion = operatorNorm(W_ges);

        if Norm_linear_subregion > theta_m
            theta_m = Norm_linear_subregion;
        end
    end
end

File: Chapter 5 Conservatism for ReLU networks/Code/Unregularized networks data generation example/ComputeCombettes.py
# function for computing the upper bound by combettes
# input  weights as needed for LipBaB, hidden_layer_dim list af dimensions of hidden layers (exclude input/output layers)

import numpy as np
import numpy.linalg as linalg
import itertools

def Theta_m(weights,hidden_layer_dim):

    matrix_list = weights

    n_hidden_layers = len(hidden_layer_dim)
    combinations_array = []

    n_hidden_neurons = np.sum(hidden_layer_dim)
    lst = list(map(list, itertools.product([0, 1], repeat=n_hidden_neurons)))
    combinations_array.append(lst)

    theta_m = 0
    for pattern in lst:
        last_value = 0
        W_ges = np.array(matrix_list[0+1])
        #print(W_ges)
        for n_layer in range(n_hidden_layers):
            Lambda = np.diag(pattern[last_value:last_value+hidden_layer_dim[n_layer]])
            #print(Lambda)
            last_value = last_value+hidden_layer_dim[n_layer]
            W_ges = np.matmul(Lambda, W_ges)
            #print(W_ges)
            W_next = np.array(matrix_list[n_layer+2])
            #print(W_next)
            W_ges = np.matmul(W_next, W_ges)
        Norm_linear_subregion = linalg.norm(W_ges,2)
        #print(Norm_linear_subregion)
        if Norm_linear_subregion > theta_m:
            theta_m = Norm_linear_subregion
    return theta_m